### ISSUE TITLE
Comments on the new code for formal stack effects

### ISSUE THREAD

**Issue by apt1002** [rrthomas/smite#71]
This issue collects my reactions to the new code to formalize stack effects for POP, DUP, SWAP, ROTATE_UP and ROTATE_DOWN.

**Comment by apt1002** [comment]
`dict.get()` can do this:

         def stack_item_to_var(item):
        -    return stack_specials.get(item) or item
        +    return stack_specials.get(item, item)

**Comment by apt1002** [comment]
I find list comprehensions easier to read than map and filter:

         # Calculate variable list corresponding to stack picture
         def stack2vars(stack):
        -    return list(map(stack_item_to_var,
        -                    filter(stack_item_has_var, stack)))
        +    return [
        +        stack_item_to_var(item)
        +        for item in stack
        +        if stack_item_has_var(item)
        +    ]

**Comment by apt1002** [comment]
You wrote:

        # Stack picture utilities
        stack_specials = {'COUNT': 'count', 'ITEMS': True}

        def stack_item_has_var(item):
            return stack_specials.get(item) != True

        def stack_item_to_var(item):
            return stack_specials.get(item, item)

I don't think there's much benefit in distinguishing "COUNT" from "count". Let's just say that the name "count" is special, and do something like this:

        COUNT_NAME = 'count'
        VARIADIC_NAME = 'ITEMS'

        def stack_item_has_var(item):
            return item != VARIADIC_NAME

        def stack_item_to_var(item):
            assert stack_item_has_var(item)
            return item

**Comment by apt1002** [comment]
In gen-actions, check_static_args only checks the static args. I can't see any check for stack underflow due to dynamic args. I think
the subsequent UNCHECKED_LOAD_STACKs could be out of bounds.

For example, here is the code generated for DUP, simplified for readability:

                smite_WORD count;
                smite_WORD dupee;
                if ((S->STACK_DEPTH < 1))
                    report underflow;
                UNCHECKED_LOAD_STACK(0, &count); // Fine.
                const smite_UWORD args = 1;
                S->STACK_DEPTH -= args;
                if ((S->STACK_SIZE - (S->STACK_DEPTH - ((smite_UWORD)count + 1)) < ((smite_UWORD)count + 1)))
                    report overflow;
                UNCHECKED_LOAD_STACK(count, &dupee); // *** Dangerous ***.
                S->STACK_DEPTH += ((smite_UWORD)count + 1) - (((smite_UWORD)count + 1) - 1);
                UNCHECKED_STORE_STACK(0, dupee);

See the line marked "*** Dangerous ***".

**Comment by apt1002** [comment]
Maybe we should represent stack pictures using an instance of StackPicture, defined below. I'm imagining this class definition would go in "vm_data.py", and I have therefore made it purpose-neutral.

This provides a place to write documentation, and a way to refer to it. It also abbreviates, clarifies or removes a lot of the logic in "gen-actions".

```
class StackPicture:
    '''
    Represents a description of the topmost items on the stack. The effect
    of an instruction can be described using two StackPictures: one for the
    arguments and one for the results.

    Variadic instructions (such as POP, DUP, SWAP) have an argument called
    "count" which indicates how many additional (unnamed) items are on the
    stack. The `is_variadic` flag is per-StackPicture, not per instruction.
    For example, POP has variadic arguments but non-variadic results.

    Public fields:
     - named_items - list of str - the names of the non-variadic items, which
       must be on the top of the stack, and might include "count".
     - is_variadic - bool - If `True`, there are `count` more items underneath
       the non-variadic items.
    '''
    def __init__(self, named_items, is_variadic=False):
        assert len(set(named_items)) == len(named_items)
        self.named_items = named_items
        self.is_variadic = is_variadic

    def static_depth(self):
        '''
        Returns the number of named stack items.
        '''
        return len(self.named_items)

    def dynamic_depth(self):
        '''
        Returns a C expression that computes the total number of items,
        including the variadic items. Assumes:
         - the C variable `count` contains the number of variadic items.
        '''
        if self.is_variadic
            return '(count + {})'.format(self.static_depth())
        else:
            return '({})'.format(self.static_depth())

```

Hmm, I see that it's not completely purpose-neutral, as it's returning C source code. A problem for later!

Update: I left it in "gen-actions" in the end, to mimimise the changes to "vm_data.py". I took the opportunity to give StackPicture some more "gen-actions"-specific methods.

**Comment by apt1002** [comment]
This utility function is useful:

```
def no_type_limits(c_source):
    '''
    Returns `c_source` wrapped in "#pragmas" to suppress warnings about some
    constant conditions, e.g. comparing an unsigned quantity to zero.
    '''
    return '''\
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wtype-limits"
{c_source}
#pragma GCC diagnostic pop
    '''.rstrip()

```

**Comment by apt1002** [comment]
The following abstract the stack limit checking:

```
def check_pop(num_pops):
    '''
    Returns C source code to check that the stack contains enough items to
    pop the specified number of items.
     - num_pops - a C expression giving the number of items to pop.
    Assumes:
     - No items have been popped so far. This is needed in the error case.
    '''
    return '''\
if (S->STACK_DEPTH < ({num_pops})) {{
    S->BAD_ADDRESS = ({num_pops});
    RAISE(3);
}}
    '''.rstrip().format(num_pops=num_pops))

def check_pop_then_push(num_pops, num_pushes):
    '''
    Returns C source code to check that the stack has enough items to pop the
    specified number of items and enough space to then push the specified
    number of items.
     - num_pops - a C expression.
     - num_pushes - a C expression.
    Assumes:
     - No items have been popped so far. This is needed in the error case.
    '''
    return '''\
{check_pop}
if (S->STACK_SIZE - (S->STACK_DEPTH - ({num_pops})) < ({num_pushes})) {{
    S->BAD_ADDRESS = ({num_pushes}) - ({num_pops});
    RAISE(2);
}}
    '''.format(
        check_pop=check_pop(num_pops),
        num_pushes=num_pushes,
    )
```

The main difference from what we have now is that these assume that `S->STACK_DEPTH` has not yet been modified, i.e. that `args` is zero. This is easy to arrange, and separates concerns.

I am imagining the sequence will be:
```
check_pop(static_arg_depth);
load_args;
check_pop_then_push(dynamic_arg_depth, dynamic_result_depth)
S->STACK_DEPTH -= ...;
do_action;
S->STACK_DEPTH += ...;
```

**Comment by apt1002** [comment]
Let's say that the modification of `S->STACK_DEPTH` is not the responsibility of `load_args()` or `store_results()`.

```
def load_args(args):
    '''
    Returns C source code to read the named arguments from the stack into C
    variables.
    `S->STACK_DEPTH` is not modified.
     - args - a StackPicture.
    '''
    return '\n'.join([
        'UNCHECKED_LOAD_STACK({pos}, &{var});'.format(pos=pos, var=var)
        for pos, var in enumerate(reversed(args.named_items))
    ])

def store_results(results):
    '''
    Returns C source code to write the named arguments from C variables into
    the stack.
    `S->STACK_DEPTH` must be modified first.
     - results - a StackPicture.
    '''
    return '\n'.join([
        'UNCHECKED_STORE_STACK({pos}, &{var});'.format(pos=pos, var=var)
        for pos, var in enumerate(reversed(results.named_items))
    ])
```

**Comment by apt1002** [comment]
We should consider rewriting the action code in "vm_data.py" to adopt the convention that `S->STACK_DEPTH` has its original value, i.e. the value before the arguments are popped. This makes them harder to read, but it has benefits:
 - It makes it easier to define `RAISE` (no need to maintain an `args` variable telling you how to fix up the stack pointer).
 - It makes it easier to write "gen-actions" (the "calling convention" of the action routines is simpler).

I'm not sure what I think about this one.

Update: I didn't do this. If it's a good idea, it should be a separate issue.

**Comment by apt1002** [comment]
Based on the above, here is the main bit of code-generation in "gen-actions":

```
    args = StackPicture.from_list(action.value.args)
    results = StackPicture.from_list(action.value.results)
    sa = args.static_depth()
    da = args.dynamic_depth()
    dr = results.dynamic_depth()
    # Concatenate the pieces.
    code = '\n'.join([
        'assert(stack_correction == 0);',
        declare_vars(args),
        declare_vars(results),
        no_type_limits(check_pop(sa)),
        load_args(args),
        no_type_limits(check_pop_then_push(da, dr)),
        'stack_correction = {}; S->STACK_DEPTH -= stack_correction;'.format(sa), # For action.value.code.
        textwrap.dedent(action.value.code.rstrip()),
        'S->STACK_DEPTH += stack_correction; stack_correction = 0;', # For action.value.code.
        'S->STACK_DEPTH -= {};'.format(da), # Effect of popping arguments.
        'S->STACK_DEPTH += {};'.format(dr), # Effect of pushing results.
        store_results(results),
    ])
```

(Note that I have renamed the C variable `args` to `stack_correction`, which more accurately expresses its purpose: the amount that `RAISE()` has to add to `S->STACK_DEPTH`).

The output looks like this (case "DUP"):

```
                assert(args == 0);
                smite_WORD count;
                smite_WORD dupee;
                #pragma GCC diagnostic push
                #pragma GCC diagnostic ignored "-Wtype-limits"
                if (S->STACK_DEPTH < (1)) {
                    S->BAD_ADDRESS = (1);
                    RAISE(3);
                }
                #pragma GCC diagnostic pop
                UNCHECKED_LOAD_STACK(0, &count);
                #pragma GCC diagnostic push
                #pragma GCC diagnostic ignored "-Wtype-limits"
                if (S->STACK_DEPTH < (((smite_UWORD)count + 1))) {
                    S->BAD_ADDRESS = (((smite_UWORD)count + 1));
                    RAISE(3);
                }
                if (S->STACK_SIZE - (S->STACK_DEPTH - (((smite_UWORD)count + 1))) < (((smite_UWORD)count + 1))) {
                    S->BAD_ADDRESS = (((smite_UWORD)count + 1)) - (((smite_UWORD)count + 1));
                    RAISE(2);
                }
                #pragma GCC diagnostic pop
                args = 1; S->STACK_DEPTH -= args;
                UNCHECKED_LOAD_STACK(count, &dupee);
                S->STACK_DEPTH += args; args = 0;
                S->STACK_DEPTH -= ((smite_UWORD)count + 1);
                S->STACK_DEPTH += ((smite_UWORD)count + 1);
                UNCHECKED_STORE_STACK(0, dupee);
```

**Comment by apt1002** [comment]
Branch "apt_stack_effect" incorporates these changes and works.

**Comment by rrthomas** [comment]
Since the structure of the code generation has changed somewhat since the version on which "apt_stack_effect" is based, I've adapted changes directly from these suggestions. I have given a thumbs up to those I've adopted, a thumbs down to those I've rejected (only one: I think STACK_DEPTH should be correct for the action code, so that code is readable when excerpted in documentation, and so that extra instructions are straightforward to write).

I have so far adapted the minor changes and the bug fix (check stack underflow for dynamic args). I shall return to the larger changes (StackPicture class and stack check generation).
